src/lib/support/crypto/RSA.cpp

```diff
@@ -0,0 +1,247 @@
+/*
+ *
+ *    Copyright (c) 2019 Google LLC.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      General RSA utility functions.
+ *
+ */
+
+#include "RSA.h"
+#include <Weave/Core/WeaveEncoding.h>
+#include <Weave/Support/CodeUtils.h>
+
+#if WEAVE_WITH_OPENSSL
+#include <openssl/bio.h>
+#include <openssl/x509.h>
+#include <openssl/rsa.h>
+#endif
+
+namespace nl {
+namespace Weave {
+namespace Crypto {
+
+using namespace nl::Weave::TLV;
+
+/**
+ * Compares with another RSA key.
+ *
+ * @param[in] other             The EncodedRSAKey object with which key should be compared.
+ *
+ * @retval true                 The keys are equal.
+ * @retval false                The keys are not equal.
+ */
+bool EncodedRSAKey::IsEqual(const EncodedRSAKey& other) const
+{
+    return (Key != NULL &&
+            other.Key != NULL &&
+            Len == other.Len &&
+            memcmp(Key, other.Key, Len) == 0);
+}
+
+/**
+ * Compares with another RSA signature.
+ *
+ * @param[in] other             The EncodedRSASignature object with which signature should be compared.
+ *
+ * @retval true                 The signatures are equal.
+ * @retval false                The signatures are not equal.
+ */
+bool EncodedRSASignature::IsEqual(const EncodedRSASignature& other) const
+{
+    return (Sig != NULL &&
+            other.Sig != NULL &&
+            Len == other.Len &&
+            memcmp(Sig, other.Sig, Len) == 0);
+}
+
+/**
+ * Reads the signature as a Weave RSASignature structure from the specified TLV reader
+ * with the given tag.
+ *
+ * @param[in] reader            The TLVReader object from which the encoded signature should
+ *                              be read.
+ * @param[in] tag               TLV tag expected to be associated with the encoded signature structure.
+ *
+ * @retval #WEAVE_NO_ERROR      If the operation succeeded.
+ * @retval other                Other Weave error codes related to signature reading.
+ */
+WEAVE_ERROR EncodedRSASignature::ReadSignature(TLVReader& reader, uint64_t tag)
```

Comment:
Passing the tag just to check it seems unnecessary.  Is this following a pattern used elsewhere?

Reply:
Agree that it is unnecessary. Updated.